File: src/430_Flatten_a_Multilevel_Doubly_Linked_List.py
# // 430. Flatten a Multilevel Doubly Linked List
# // https://leetcode.cn/problems/flatten-a-multilevel-doubly-linked-list/



# Definition for a Node.
class Node:
    def __init__(self, val, prev, next, child):
        self.val = val
        self.prev = prev
        self.next = next
        self.child = child

class Solution:
    def flatten(self, head: "Node") -> "Node":
        def dfs(node: "Node") -> "Node":
            cur = node
            # 记录链表的最后一个节点
            last = None

            while cur:
                nxt = cur.next
                # 如果有子节点，那么首先处理子节点
                if cur.child:
                    child_last = dfs(cur.child)
                    
                    nxt = cur.next
                    # 将 node 与 child 相连
                    cur.next = cur.child
                    cur.child.prev = cur

                    # 如果 nxt 不为空，就将 last 与 nxt 相连
                    if nxt:
                        child_last.next = nxt
                        nxt.prev = child_last

                    # 将 child 置为空
                    cur.child = None
                    last = child_last
                else:
                    last = cur
                cur = nxt

            return last
        
        dfs(head)
        return head

File: src/430_Flatten_a_Multilevel_Doubly_Linked_List.java
// // https://leetcode.cn/problems/flatten-a-multilevel-doubly-linked-list/

// // Definition for a Node.
// class Node {
//     public int val;
//     public Node prev;
//     public Node next;
//     public Node child;
// };



// class Solution {
//     public Node flatten(Node head) {
//         dfs(head);
//         return head;
//     }
//     Node dfs(Node head) {
//         Node last = head;
//         while (head != null) {
//             if (head.child == null) {
//                 last = head;
//                 head = head.next;
//             } else {
//                 Node tmp = head.next;
//                 Node childLast = dfs(head.child);
//                 head.next = head.child;
//                 head.child.prev = head;
//                 head.child = null;
//                 if (childLast != null) childLast.next = tmp;
//                 if (tmp != null) tmp.prev = childLast;
//                 last = head;
//                 head = childLast;
//             }
//         }
//         return last;
//     }
// }

// 作者：AC_OIer
// 链接：https://leetcode.cn/problems/flatten-a-multilevel-doubly-linked-list/solution/gong-shui-san-xie-yi-ti-shuang-jie-di-gu-9wfz/
// 来源：力扣（LeetCode）
// 著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。


// class Solution {
//     public Node flatten(Node head) {
//         Node dummy = new Node(0);
//         dummy.next = head;

//         while (head != null) {
//             if (head.child == null) {
//                 head = head.next;
//             } else {
//                 Node tmp = head.next;
//                 Node child_head = flatten(head.child);
//                 head.next = child_head;
//                 child_head.prev = head;
//                 head.child = null;
//                 while (head.next != null) { head = head.next; }
//                 head.next = tmp;
//                 if (tmp != null) { tmp.prev = head; }
//                 head = tmp;
//             }
//         }

//         return dummy.next;
//     }
// }

// class Solution {
//     public Node flatten(Node head) {
//         Node dummy = new Node(0);
//         dummy.next = head;

//         for (; head != null; head = head.next) {
//             if (head.child != null) {
//                 Node tmp = head.next;
//                 Node child = head.child;
//                 head.next = child;
//                 child.prev = head;
//                 head.child = null;
//                 Node last = head;
//                 while (last.next != null) last = last.next;
//                 last.next = tmp;
//                 if (tmp != null) tmp.prev = last;
//             }
//         }
//         return dummy.next;
//     }
// }

// // 作者：AC_OIer
// // 链接：https://leetcode.cn/problems/flatten-a-multilevel-doubly-linked-list/solution/gong-shui-san-xie-yi-ti-shuang-jie-di-gu-9wfz/
// // 来源：力扣（LeetCode）
// // 著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。